packages/boot/src/boot.mixin.ts

```diff
@@ -0,0 +1,129 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/boot
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor, Binding, BindingScope, Context} from '@loopback/context';
+import {Booter, BootOptions, Bootable} from './interfaces';
+import {BootComponent} from './boot.component';
+import {Bootstrapper} from './bootstrapper';
+import {BootBindings} from './keys';
+
+// Binding is re-exported as Binding / Booter types are needed when consuming
+// BootMixin and this allows a user to import them from the same package (UX!)
+export {Binding};
+
+/**
+ * Mixin for @loopback/boot. This Mixin provides the following:
+ * - Implements the Bootable Interface as follows.
+ * - Add a `projectRoot` property to the Class
+ * - Adds an optional `bootOptions` property to the Class that can be used to
+ *    store the Booter conventions.
+ * - Adds the `BootComponent` to the Class (which binds the Bootstrapper and default Booters)
+ * - Provides the `boot()` convenience method to call Bootstrapper.boot()
+ * - Provides the `booter()` convenience method to bind a Booter(s) to the Application
+ * - Override `component()` to call `mountComponentBooters`
+ * - Adds `mountComponentBooters` which binds Booters to the application from `component.booters[]`
+ */
+// tslint:disable-next-line:no-any
+export function BootMixin<T extends Constructor<any>>(superClass: T) {
+  return class extends superClass implements Bootable {
+    projectRoot: string;
+    bootOptions?: BootOptions;
+
+    // tslint:disable-next-line:no-any
+    constructor(...args: any[]) {
+      super(...args);
+      this.component(BootComponent);
+    }
+
+    /**
+     * Convenience method to call bootstrapper.boot()
+     */
+    async boot(): Promise<void> {
+      // Project Root and Boot Options need to be bound so we can resolve an
+      // instance of Bootstrapper.
+      this.bind(BootBindings.PROJECT_ROOT).to(this.projectRoot);
+      this.bind(BootBindings.BOOT_OPTIONS).to(this.bootOptions);
+
+      // Get a instance of the BootStrapper
+      const bootstrapper: Bootstrapper = await this.get(
+        BootBindings.BOOTSTRAPPER_KEY,
+      );
+
+      await bootstrapper.boot();
+    }
+
+    /**
+     * Given a N number of Booter Classes, this method binds them using the
+     * prefix and tag expected by the Bootstrapper.
+     *
+     * @param booterCls Booter classes to bind to the Application
+     *
+     * ```ts
+     * app.booter(MyBooter, MyOtherBooter)
+     * ```
+     */
+    booter(...booterCls: Constructor<Booter>[]): Binding[] {
+      // tslint:disable-next-line:no-any
+      return booterCls.map(cls => _bindBooter(<Context>(<any>this), cls));
```

Comment:
`map` is used to transform an array of X values to an array of Y values. Please use `forEach` instead.

I find this typecast very worrying: `<Context>(<any>this)`. Isn't there any way how to tell TypeScript that `BootMixin` can be applied only on classes inheriting from `Context`?

Reply:
`map` is used because we're transforming an array of `Constructor<Booter>` classes to an array of `Binding`. Using `forEach` will mean manually adding bindings to an array to return.

I had tried constraining the Mixin type to `Application` but it caused issues with TypeScript ... issues relating to requiring me to import every definition used by Application into the file where we use the Mixin ... OR ... issues with being able to chain mixins. I'll add comments documenting this as per @raymondfeng's suggestion for future reference and maybe at some point there will be a way to fix this.